### PR TITLE
[WIP] fix for the slow updates of targets changes

### PR DESCRIPTION
**Pull request by krasi-georgiev**
The scrape manage receiver channel now just saves the target sets
and another backgorund runner updates the scrape loops every 5 seconds.
This is so that the scrape manager doesn't block the receiving channel
when it does the  long background reloading of the scrape loops.

Active and dropped targets are now saved in each scrape pool instead of
the scrape manager. This is mainly to avoid races when getting the
targets via the web api.

When reloading the scrape loops now happens in parallel to speed up the
final disared state and this also speeds up the prometheus's shutting
down.

fixes: https://github.com/prometheus/prometheus/issues/4124
fixes: https://github.com/prometheus/prometheus/issues/4301

I will run a benchmark once the https://github.com/prometheus/prometheus/pull/4523 is merged and I rebase it to this PR as well.

Signed-off-by: Krasi Georgiev <kgeorgie@redhat.com>

### CODE REVIEW

**Comment by krasi-georgiev** [comment]
@grobie still polishing few details, but you can already test this version or wait few days to clear all small issues.

The last major problem is with the http2 go client update in https://github.com/golang/net/commit/1c05540f6879653db88113bc4a2b70aec4bd491f , but as long at your k8s cluster doesn't use the default connection stream limit of 1000, you should be fine.
Alternatevely you can disable HTTP2 by running prometheus with 
```
DISABLE_HTTP2=true ./prometheus ....
```

more details about the http2 client issue.
https://github.com/kubernetes/client-go/issues/456

**Comment by krasi-georgiev** [comment]
the benchmark is looking good. When I scale up and down the CPU and memory usage is higher, but this is to be expected as all processing is now done in parallel.

File: scrape/manager.go [hunk 1]
```diff
@@ -50,28 +51,84 @@ type Manager struct {
...
+	m.mtxScrape.Lock()
+	var wg sync.WaitGroup
+	for setName, groups := range m.targetSets {
+		var sp *scrapePool
+		existing, ok := m.scrapePools[setName]
+		if !ok {
+			scrapeConfig, ok := m.scrapeConfigs[setName]
+			if !ok {
+				level.Error(m.logger).Log("msg", "error reloading target set", "err", fmt.Sprintf("invalid config id:%v", setName))
+				return
```
Comment (simonpasquier) [inline scrape/manager.go]: Could it happen? Does it need to return or it could continue?

Right now the scrape manager's configuration is reloaded before the discovery manager.

https://github.com/prometheus/prometheus/blob/75bd348135019c6bf5bee4350a339964e62f2d8a/cmd/prometheus/main.go#L323-L333

So I guess it can happen when a scrape job is removed from the configuration and the configuration update has been applied to the scrape manager but the discovery manager hasn't picked up yet. I'd say that we can silently drop the item then.

_I see that this code already exists in the previous version so feel free to ignore_

Comment (krasi-georgiev) [reply scrape/manager.go]: If someone changes the workflow somewhere we might hit this so this way we will know soon enough that something is not right. This is one of the problems with the current implementation of how wet handle the config reloading. 
I have suggested how to improve, but there was no feedback at the time.

Comment (krasi-georgiev) [reply scrape/manager.go]: Either way this is story for another PR ;)

**Comment by beorn7** [comment]
Hi @krasi-georgiev , I'm back at work and would like to canary all those SD changes. What's the state of this PR? Should this still be included? You said you wanted to rebase this before further checking.

**Comment by simonpasquier** [comment]
@beorn7 IIUC #4124 that you reported on targets not being updated should be fixed now on master by https://github.com/prometheus/prometheus/pull/4523 and  https://github.com/prometheus/prometheus/pull/4556. This PR is more about optimizations that would speed up the recreation of the scrape loops. Also #3912 (which merges identical SD configurations into a single provider and thus avoids hammering the K8S API) has been merged. Finally Krasi has found a couple of issues with the K8S go client (#4518 and #4528) but not sure that they apply to your environment.

In summary, you're more than welcome to test the current master on your clusters :smirk:

**Comment by krasi-georgiev** [comment]
@beorn7 , @grobie would you mind testing this?

**Comment by simonpasquier** [comment]
LGTM but I think it would worth to have unit tests for this.

**Comment by krasi-georgiev** [comment]
@simonpasquier what unit tests did you have in mind? 

This PR fixes the time it takes to reload the scrape loops. 
An e2e with some timings  would be flaky, but suggestions are welcome.

**Comment by beorn7** [comment]
I can run this on one of our servers at my next convenience.

**Comment by krasi-georgiev** [comment]
@beorn7 appreciated!

**Comment by simonpasquier** [comment]
I was thinking about something similar to https://github.com/prometheus/prometheus/pull/4582 although looking at the manager's code it would require some adjustments to be able to skip the effective launch of the scrapers.

**Comment by beorn7** [comment]
Running this now (with current state of master merged in).

Disclaimer: I'm not using race detector (would not work on the heavily loaded test machine), but I recommend to do so before merging this.

I guess this change should result in faster reloads of the config file. I sighup'd the test server and the 2.4.0 production server a few times and couldn't detect a significant improvement, mostly because reload times are vastly different from case to case, from 5s to 30s, more or less the same on both servers.

Restart time is quite long on this server (around 12m (!)) and isn't significantly different with this PR merged in. (Perhaps the new WAL implementation is slower? I don't remember such long startup times from earlier 2.x versions. But perhaps our load has just increased so much in the meantime...)

Altogether, the two servers seem to behave mostly the same.

**Comment by krasi-georgiev** [comment]
I tested this mainly with the k8s discovery.
Using the config from: https://github.com/prometheus/prometheus/issues/4124#issuecomment-412441386

in general 100 jobs with 20-30 targets per jobs.

Than randomly scaling up and down the number of targets from 1 to 20 .

Before this pr it takes minutes to reflect the changes in the  /targets web gui and it can never catch up.

**Comment by krasi-georgiev** [comment]
When I tested locally it also improved the shutdown time as now old scrape loops are stopped in parallel.

**Comment by simonpasquier** [comment]
@krasi-georgiev have you retried your tests with the latest master? I suspect that the various improvements to the SD manager may have change the situation.

**Comment by beorn7** [comment]
@krasi-georgiev I see. I guess our test server doesn't see quickly changing targets that often. I might be able to identify one of those, and then run the canary binary there.

**Comment by beorn7** [comment]
Looking around, with 2.4.0, all our practical uses of K8s SD seem to update fast enough so that any further improvements would not stick out of the noise.

**Comment by krasi-georgiev** [comment]
Maybe @brancz can also give it a try since it is very easy to replicate using his configs even using minikibe.

@simonpasquier this change is in the scrape manager which is completely independent from the sd manger. 
Which changes do you think might be relevant here?

**Comment by simonpasquier** [comment]
>  Which changes do you think might be relevant here

Mostly #3912. It might be that optimizing the discovery part (basically much less requests to the K8S API) reduced the contention and made the problem less acute on the scrape side.

**Comment by krasi-georgiev** [comment]
aah yes that definitely helped, but processing the loops in parallel is another great improvement.

**Comment by simonpasquier** [comment]
Scaling from 0 to 2000 targets back and forth, I definitely see an improvement when down-scaling. WDYT of adding a few tests?

**Comment by krasi-georgiev** [comment]
the only useful test that I can see is making  sure the Scrape manager never blocks the `tsets <-chan map[string][]*targetgroup.Group` 

Is this what you had in mind as well?

**Comment by simonpasquier** [comment]
Yes. Also making sure that data received on the channel triggers scrape updates and that everything is teared down properly on shutdown. I know that it wasn't tested before but it looks like a good opportunity to fix it.

**Comment by krasi-georgiev** [comment]
@simonpasquier yes good idea, I added some tests. 

Just not sure what you had in mind for: 

> everything is teared down properly on shutdown.

File: scrape/manager.go [hunk 2]
```diff
@@ -50,28 +54,83 @@ type Manager struct {
...
+			}
+			sp = newScrapePool(scrapeConfig, m.append, log.With(m.logger, "scrape_pool", setName))
+			m.scrapePools[setName] = sp
+		} else {
+			sp = existing
+		}
+
+		wg.Add(1)
+		// Run the sync in paralel as these take a while and at high load can't catch up.
```
Comment (simonpasquier) [inline scrape/manager.go]: s/paralel/parallel/

**Comment by simonpasquier** [comment]
:+1: It would be even better to test the logic including `reload()` but it is more involved and wasn't covered before either... 

> everything is teared down properly on shutdown.

That calling `Stop()` does the right thing.

**Comment by krasi-georgiev** [comment]
Yeah I didn't want to take it that far as this would probably end being too high maintenance. 
In general it is not a bad idea, but it is outside the topic of this PR.
Even if someone decides to implement it let's leave it for another PR so we keep this one on topic and don't block it for any longer that is needed. 
The same for the teardown.
